### PR TITLE
add possibility to disable dxc

### PR DESCRIPTION
**Pull request by AndreyMlashkin**
Some users may not need DirectXShaderCompiler support.
Introduce a cmake flag to disable it

### CODE REVIEW

**Comment by TheMostDiligent** [comment]
Please rename this option to `DILIGENT_NO_DXC` and move it to the ShaderTools' CMake file.

**Comment by TheMostDiligent** [comment]
Why do you need to disable it, by the way?

**Comment by AndreyMlashkin** [comment]
I need to run diligent in a docker container and I need to keep it as small as possible.

**Comment by AndreyMlashkin** [comment]
Is there a CI problem?
By default, nothing should changed, only if user overrides the flag

**Comment by azhirnov** [comment]
@AndreyMlashkin DirectX shaders uses DXC to get reflection from DXIL, you should surround this code by macros

**Comment by TheMostDiligent** [comment]
> Is there a CI problem?

Yes, DX12 backend expects that DXC is always present and there is a build error. It is expected to be the main compiler for DX12. If you disable it, you will not be able to consume DXIL. None of the next-gen DX12 features (mesh shaders, ray tracing, wave operations, etc.) will not work.

> I need to run diligent in a docker container and I need to keep it as small as possible.

You are disabling one file worth of ~1800 lines of code. Not sure this will give you a ton of space savings.
Note that you are not disabling DXC itself, but only the code that calls. The compiler itself is in the System directory or the SDK.

**Comment by AndreyMlashkin** [comment]
I use vulkan installation from conan, it contains only minimal libs (only vulkanlib actually), there is no dxc compiler at all.
That's why I also want to exclude code, that uses it

File: Graphics/ShaderTools/CMakeLists.txt [hunk 1]
```diff
@@ -39,7 +41,7 @@ if(D3D11_SUPPORTED OR D3D12_SUPPORTED)
     list(APPEND INCLUDE include/ResourceBindingMap.hpp)
 endif()
 
-if((PLATFORM_WIN32 AND NOT MINGW_BUILD) OR PLATFORM_UNIVERSAL_WINDOWS OR PLATFORM_LINUX)
+if(DILIGENT_NO_DXC AND ((PLATFORM_WIN32 AND NOT MINGW_BUILD) OR PLATFORM_UNIVERSAL_WINDOWS OR PLATFORM_LINUX))
```
Comment (TheMostDiligent) [inline Graphics/ShaderTools/CMakeLists.txt:44]: Should be `NOT DILIGENT_NO_DXC`

File: Graphics/ShaderTools/CMakeLists.txt [hunk 2]
```diff
@@ -2,6 +2,8 @@ cmake_minimum_required (VERSION 3.6)
 
 project(Diligent-ShaderTools CXX)
 
+set(DILIGENT_NO_DXC FALSE CACHE INTERNAL "Use DirectXShaderCompiler compiler")
```
Comment (TheMostDiligent) [inline Graphics/ShaderTools/CMakeLists.txt:5]: I don't think this needs to be a cache variable. If it is internal and not an CMake option, there is no need to define this variable here at all.

Comment (TheMostDiligent) [reply Graphics/ShaderTools/CMakeLists.txt:5]: Besides, this variable defines whether to **disable** using the compiler, not when to use it

Comment (TheMostDiligent) [reply Graphics/ShaderTools/CMakeLists.txt:5]: Please remove this variable, it is not needed

**Review by TheMostDiligent — CHANGES_REQUESTED** [review]

**Comment by AndreyMlashkin** [comment]
I have tried to compile my changes only in Windows for different flags.

**Comment by AndreyMlashkin** [comment]
@TheMostDiligent , I would kindly ask you to start the CI pipeline

File: Graphics/ShaderTools/CMakeLists.txt [hunk 3]
```diff
@@ -43,7 +43,11 @@ if((PLATFORM_WIN32 AND NOT MINGW_BUILD) OR PLATFORM_UNIVERSAL_WINDOWS OR PLATFOR
     set(DXC_SUPPORTED TRUE)
 endif()
 
-if (DXC_SUPPORTED)
+if(NOT DILIGENT_NO_DIRECT3D11 OR NOT DILIGENT_NO_DIRECT3D12)
+  set(DXC_NEEDED TRUE)
+endif()
```
Comment (TheMostDiligent) [inline Graphics/ShaderTools/CMakeLists.txt:48]: This logic is incorrect. On Linux, for example, there is no D3D11 and D3D12 even when neither DILIGENT_NO_DIRECT3D11 nor DILIGENT_NO_DIRECT3D12 is defined, but DXC is still needed.
The correct check is if (D3D11_SUPPORTED or D3D12_SUPPORTED)

File: Graphics/ShaderTools/CMakeLists.txt [hunk 4]
```diff
@@ -43,7 +43,11 @@ if((PLATFORM_WIN32 AND NOT MINGW_BUILD) OR PLATFORM_UNIVERSAL_WINDOWS OR PLATFOR
...
+
+if (DXC_SUPPORTED AND DXC_NEEDED)
```
Comment (TheMostDiligent) [inline Graphics/ShaderTools/CMakeLists.txt:50]: This is also incorrect. DXC is used by Vulkan backend even when there is no D3D11 and D3D12. All high-end tutorials (mesh shaders, ray tracing, hybrid rendering) require DXC compiler in Vulkan mode

**Comment by TheMostDiligent** [comment]
I am leaning to close this PR as it does not really provide any benefit. Removing one file from the build is not going to give any significant space savings, but complicates the build configuration by a lot. If you don't have DXC on your system, everything will still run fine unless you try to use `SHADER_COMPILER_DXC`. You need to install ATL support for Visual Studio and the file will compile without errors.

**Pull request closed without merging**